### ISSUE TITLE
Source location is sometimes a list

### ISSUE THREAD

**Issue by jc00ke** [elixir-lang/ex_doc#580]
With `ex_doc 1.2.0` on `Elixir 1.3.1` and `Erlang 19` I see the following error (which I did not used to see on earlier versions)

```
➜ cerebro (sync-docs-built-from-CI) ✗ mix docs
==> ex_doc
Compiling 5 files (.ex)
==> cerebro
Compiling 33 files (.ex)
Generated cerebro app
** (ArgumentError) cannot convert the given list to a string.

To be converted to a string, a list must contain only:

  * strings
  * integers representing Unicode codepoints
  * or a list containing one of these three elements

Please check the given list or call inspect/1 to get the list representation, got:

[location: 1, generated: true]

    (elixir) lib/list.ex:624: List.to_string/1
    lib/ex_doc/retriever.ex:427: ExDoc.Retriever.source_link/3
    lib/ex_doc/retriever.ex:206: ExDoc.Retriever.get_function/6
    lib/ex_doc/retriever.ex:156: anonymous fn/8 in ExDoc.Retriever.get_docs/7
    (elixir) lib/enum.ex:1623: Enum."-reduce/3-lists^foldl/2-0-"/3
    lib/ex_doc/retriever.ex:155: ExDoc.Retriever.get_docs/7
    lib/ex_doc/retriever.ex:123: ExDoc.Retriever.generate_node/3
    (elixir) lib/enum.ex:1184: Enum."-map/2-lists^map/1-0-"/2
    (elixir) lib/enum.ex:1184: Enum."-map/2-lists^map/1-0-"/2
    lib/ex_doc/retriever.ex:69: ExDoc.Retriever.docs_from_modules/2
    lib/ex_doc.ex:48: ExDoc.generate_docs/3
    lib/mix/tasks/docs.ex:119: Mix.Tasks.Docs.run/3
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
    (elixir) lib/enum.ex:651: Enum.each/2
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
```

A quick patch of

```patch
diff --git a/lib/ex_doc/retriever.ex b/lib/ex_doc/retriever.ex
index 11dcb20..14074e6 100644
--- a/lib/ex_doc/retriever.ex
+++ b/lib/ex_doc/retriever.ex
@@ -409,9 +409,12 @@ defmodule ExDoc.Retriever do
 
   defp source_link(source_path, source_url, line) do
     source_url = Regex.replace(~r/%{path}/, source_url, source_path)
-    Regex.replace(~r/%{line}/, source_url, to_string(line))
+    Regex.replace(~r/%{line}/, source_url, source_line(line))
   end
 
+  defp source_line(line) when is_list(line), do: to_string(line[:location])
+  defp source_line(line), do: to_string(line)
+
   defp source_path(module, config) do
     source = module.__info__(:compile)[:source] |> String.Chars.to_string
 ```

does the trick and all tests pass, and my docs compile. I'm having a hard time tracking down exactly what is causing this to break in my (closed-source) app though.

I tried `rescue`'ing the exception and tracking down the source of the issue but I didn't get far.

I'd love help figuring this out/making an acceptable PR (with tests!)

Thanks!

**Comment by josevalim** [comment]
Do we have 0.12.1 out? If not, can you try master? I remember fixing a
remotely similar problem.

On Thursday, July 14, 2016, Jesse Cooke <notifications@github.com> wrote:

> With ex_doc 1.2.0 on Elixir 1.3.1 and Erlang 19 I see the following error
> (which I did not used to see on earlier versions)
>
> ➜ cerebro (sync-docs-built-from-CI) ✗ mix docs
> ==> ex_doc
> Compiling 5 files (.ex)
> ==> cerebro
> Compiling 33 files (.ex)
> Generated cerebro app
> ** (ArgumentError) cannot convert the given list to a string.
>
> To be converted to a string, a list must contain only:
>
>   * strings
>   * integers representing Unicode codepoints
>   * or a list containing one of these three elements
>
> Please check the given list or call inspect/1 to get the list representation, got:
>
> [location: 1, generated: true]
>
>     (elixir) lib/list.ex:624: List.to_string/1
>     lib/ex_doc/retriever.ex:427: ExDoc.Retriever.source_link/3
>     lib/ex_doc/retriever.ex:206: ExDoc.Retriever.get_function/6
>     lib/ex_doc/retriever.ex:156: anonymous fn/8 in ExDoc.Retriever.get_docs/7
>     (elixir) lib/enum.ex:1623: Enum."-reduce/3-lists^foldl/2-0-"/3
>     lib/ex_doc/retriever.ex:155: ExDoc.Retriever.get_docs/7
>     lib/ex_doc/retriever.ex:123: ExDoc.Retriever.generate_node/3
>     (elixir) lib/enum.ex:1184: Enum."-map/2-lists^map/1-0-"/2
>     (elixir) lib/enum.ex:1184: Enum."-map/2-lists^map/1-0-"/2
>     lib/ex_doc/retriever.ex:69: ExDoc.Retriever.docs_from_modules/2
>     lib/ex_doc.ex:48: ExDoc.generate_docs/3
>     lib/mix/tasks/docs.ex:119: Mix.Tasks.Docs.run/3
>     (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
>     (elixir) lib/enum.ex:651: Enum."-each/2-lists^foreach/1-0-"/2
>     (elixir) lib/enum.ex:651: Enum.each/2
>     (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
>     (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
>
> A quick patch of
>
> diff --git a/lib/ex_doc/retriever.ex b/lib/ex_doc/retriever.ex
> index 11dcb20..14074e6 100644--- a/lib/ex_doc/retriever.ex+++ b/lib/ex_doc/retriever.ex@@ -409,9 +409,12 @@ defmodule ExDoc.Retriever do
>
>    defp source_link(source_path, source_url, line) do
>      source_url = Regex.replace(~r/%{path}/, source_url, source_path)-    Regex.replace(~r/%{line}/, source_url, to_string(line))+    Regex.replace(~r/%{line}/, source_url, source_line(line))
>    end
> +  defp source_line(line) when is_list(line), do: to_string(line[:location])+  defp source_line(line), do: to_string(line)+
>    defp source_path(module, config) do
>      source = module.__info__(:compile)[:source] |> String.Chars.to_string
>
> does the trick and all tests pass, and my docs compile. I'm having a hard
> time tracking down exactly what is causing this to break in my
> (closed-source) app though.
>
> I tried rescue'ing the exception and tracking down the source of the
> issue but I didn't get far.
>
> I'd love help figuring this out/making an acceptable PR (with tests!)
>
> Thanks!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/ex_doc/issues/580>, or mute the thread
> <https://github.com/notifications/unsubscribe/AAAlbn17z3HJa74xFjgXWmU-QR7ZXePVks5qVn7bgaJpZM4JMtkD>
> .
>


-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

**Comment by jc00ke** [comment]
`0.12.1` has not been released yet, and `master` still :boom: 

```
➜ cerebro (sync-docs-built-from-CI) ✗ mix docs
==> earmark
Compiling 1 file (.erl)
Compiling 16 files (.ex)
warning: variable pending_lnb is unused
  lib/earmark/helpers/lookahead_helpers.ex:34

warning: unused import Earmark.Helpers.StringHelpers
  lib/earmark/helpers/lookahead_helpers.ex:7

Generated earmark app
==> ex_doc
Compiling 12 files (.ex)
Generated ex_doc app
==> cerebro
Compiling 33 files (.ex)
Generated cerebro app
** (ArgumentError) cannot convert the given list to a string.

To be converted to a string, a list must contain only:

  * strings
  * integers representing Unicode codepoints
  * or a list containing one of these three elements

Please check the given list or call inspect/1 to get the list representation, got:

[location: 1, generated: true]

    (elixir) lib/list.ex:624: List.to_string/1
    lib/ex_doc/retriever.ex:412: ExDoc.Retriever.source_link/3
    lib/ex_doc/retriever.ex:206: ExDoc.Retriever.get_function/6
    lib/ex_doc/retriever.ex:156: anonymous fn/8 in ExDoc.Retriever.get_docs/7
    (elixir) lib/enum.ex:1623: Enum."-reduce/3-lists^foldl/2-0-"/3
    lib/ex_doc/retriever.ex:155: ExDoc.Retriever.get_docs/7
    lib/ex_doc/retriever.ex:123: ExDoc.Retriever.generate_node/3
    (elixir) lib/enum.ex:1184: Enum."-map/2-lists^map/1-0-"/2
    (elixir) lib/enum.ex:1184: Enum."-map/2-lists^map/1-0-"/2
    lib/ex_doc/retriever.ex:69: ExDoc.Retriever.docs_from_modules/2
    lib/ex_doc.ex:48: ExDoc.generate_docs/3
    lib/mix/tasks/docs.ex:119: Mix.Tasks.Docs.run/3
    (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:363: Code.require_file/2
```

That's with the latest `Earmark` too.

**Comment by josevalim** [comment]
Cool. I think you find a bug in the Elixir compiler. Do you have a sample
app that triggers it? :)

On Thursday, July 14, 2016, Jesse Cooke <notifications@github.com> wrote:

> 0.12.1 has not been released yet, and master still 💥
>
> ➜ cerebro (sync-docs-built-from-CI) ✗ mix docs
> ==> earmark
> Compiling 1 file (.erl)
> Compiling 16 files (.ex)
> warning: variable pending_lnb is unused
>   lib/earmark/helpers/lookahead_helpers.ex:34
>
> warning: unused import Earmark.Helpers.StringHelpers
>   lib/earmark/helpers/lookahead_helpers.ex:7
>
> Generated earmark app
> ==> ex_doc
> Compiling 12 files (.ex)
> Generated ex_doc app
> ==> cerebro
> Compiling 33 files (.ex)
> Generated cerebro app
> ** (ArgumentError) cannot convert the given list to a string.
>
> To be converted to a string, a list must contain only:
>
>   * strings
>   * integers representing Unicode codepoints
>   * or a list containing one of these three elements
>
> Please check the given list or call inspect/1 to get the list representation, got:
>
> [location: 1, generated: true]
>
>     (elixir) lib/list.ex:624: List.to_string/1
>     lib/ex_doc/retriever.ex:412: ExDoc.Retriever.source_link/3
>     lib/ex_doc/retriever.ex:206: ExDoc.Retriever.get_function/6
>     lib/ex_doc/retriever.ex:156: anonymous fn/8 in ExDoc.Retriever.get_docs/7
>     (elixir) lib/enum.ex:1623: Enum."-reduce/3-lists^foldl/2-0-"/3
>     lib/ex_doc/retriever.ex:155: ExDoc.Retriever.get_docs/7
>     lib/ex_doc/retriever.ex:123: ExDoc.Retriever.generate_node/3
>     (elixir) lib/enum.ex:1184: Enum."-map/2-lists^map/1-0-"/2
>     (elixir) lib/enum.ex:1184: Enum."-map/2-lists^map/1-0-"/2
>     lib/ex_doc/retriever.ex:69: ExDoc.Retriever.docs_from_modules/2
>     lib/ex_doc.ex:48: ExDoc.generate_docs/3
>     lib/mix/tasks/docs.ex:119: Mix.Tasks.Docs.run/3
>     (mix) lib/mix/task.ex:296: Mix.Task.run_task/3
>     (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
>     (elixir) lib/code.ex:363: Code.require_file/2
>
> That's with the latest Earmark too.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/ex_doc/issues/580#issuecomment-232759009>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAlbu6X9BCFcJWzYVTdp7ncRHjHRcQzks5qVoZAgaJpZM4JMtkD>
> .
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by jc00ke** [comment]
I'll try to reduce my app to a minimum.

**Comment by josevalim** [comment]
@jc00ke thank you! fwiw, it is definitely something macro related.

**Comment by jc00ke** [comment]
Still working on tracking it down. Having a hard time making a concise repro, but my app does consistently error out on `:__phoenix_recompile__?/0` with `source_path` of `"web/views/error_view.ex"`.

**Comment by josevalim** [comment]
@jc00ke if you start a new phoenix project and try to build docs for it, does it work or fail?

**Comment by jc00ke** [comment]
It works, unfortunately. So I'm adding in dependencies to see if one of them triggers it, slowly building back up.

**Comment by jc00ke** [comment]
Found it! I had missed setting the `source_url` in `mix.exs`.
Here's the repro repo: https://github.com/jc00ke/ex_doc_example

**Comment by josevalim** [comment]
@jc00ke thank you, i have pushed a fix to master. Once confirmed, I will do a new release!

**Comment by jc00ke** [comment]
Confirmed! Thanks for such a quick fix!

closed by 1381d861801c703cf37df885938abe21e8340544

**Comment by josevalim** [comment]
New version is out!
